src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -8380,6 +8380,62 @@ internal static string HandleDoubleAndSingleQuote(ref string wordToComplete)
             return quote;
         }
 
+        /// <summary>
+        /// Gets matching results from word to complete.
+        /// This makes it easier to handle different variations of completions with consideration of quotes.
+        /// </summary>
+        /// <param name="wordToComplete">The word to complete.</param>
+        /// <param name="possibleCompletionValues">The possible completion values to iterate.</param>
+        /// <param name="toolTipMapping">The optional tool tip mapping delegate.</param>
+        /// <param name="resultType">The optional completion result type. Default is Text.</param>
+        /// <returns></returns>
+        internal static IEnumerable<CompletionResult> GetMatchingResults(
+            string wordToComplete,
+            IEnumerable<string> possibleCompletionValues,
+            Func<string, string> toolTipMapping = null,
+            CompletionResultType resultType = CompletionResultType.Text)
+        {
+            string quote = HandleDoubleAndSingleQuote(ref wordToComplete);
+            var pattern = WildcardPattern.Get(wordToComplete + "*", WildcardOptions.IgnoreCase);
+
+            foreach (string value in possibleCompletionValues)
+            {
+                string completionText = value;
+                string listItemText = value;
+
+                if (pattern.IsMatch(completionText))
+                {
+                    if (CompletionRequiresQuotes(completionText, escape: false))
```

Comment:
Most code is not setting `escape` here for `CompletionRequiresQuotes` so assuming we don't need to care for `[` or `]` for now. I am currently handling single quote, backtick and `$` escaping.